lib/services.js

```diff
@@ -288,18 +296,21 @@ export default {
      * @param {string} objectKey - object key name
      * @param {Log} log - logger instance
      * @param {function} cb - callback from async.waterfall in objectGet
+     * @param {object} query - query of the request
      * @return {undefined}
      */
-    deleteObject(bucketName, objectMD, responseMDHeaders, objectKey, log, cb) {
+    deleteObject(bucketName, objectMD, responseMDHeaders, objectKey, log, cb,
+        query) {
         log.trace('deleting object from bucket');
         assert.strictEqual(typeof bucketName, 'string');
         assert.strictEqual(typeof objectMD, 'object');
-        if (objectMD['x-amz-version-id'] === 'null') {
+        if (objectMD['x-amz-version-id'] === 'null' || query.versionId) {
```

Comment:
No it's part of the opposite. This check is to see if we are deleting a non-versioned object or a specific version---cases when we need to physically delete object's data. The later part of the condition means when the object is versioning-enabled and the request is to delete a specific version; it is equivalent to `objectMD['x-amz-version-id'] !== 'null' && query.versionId`.